go/client/logclient.go

```diff
@@ -314,11 +371,124 @@ func (c *LogClient) GetSTH() (sth *ct.SignedTreeHead, err error) {
 	if err != nil {
 		return nil, err
 	}
-	// TODO(alcutter): Verify signature
 	sth.TreeHeadSignature = *ds
+	err = c.verifySTH(sth)
+	if err != nil {
+		return nil, err
+	}
 	return
 }
 
+func (c *LogClient) verifySignature(data []byte, signed ct.DigitallySigned) error {
+	// Only support SHA-256 for the hash function.
+	if signed.HashAlgorithm != ct.SHA256 {
+		return fmt.Errorf("Unexpected hash algorithm %v", signed.HashAlgorithm)
+	}
+	hash := sha256.Sum256(data)
+
+	switch pubkey := c.pubkey.(type) {
+	case *rsa.PublicKey:
+		if signed.SignatureAlgorithm != ct.RSA {
+			return fmt.Errorf("Unexpected signature algorithm %v", signed.SignatureAlgorithm)
+		}
+		if err := rsa.VerifyPKCS1v15(pubkey, crypto.SHA256, hash[:], signed.Signature); err != nil {
+			return fmt.Errorf("RSA signature validation failed %v", err.Error())
+		}
+		return nil
+	case *ecdsa.PublicKey:
+		if signed.SignatureAlgorithm != ct.ECDSA {
+			return fmt.Errorf("Unexpected signature algorithm %v", signed.SignatureAlgorithm)
+		}
+		var rs struct {
+			R, S *big.Int
+		}
+		if _, err := asn1.Unmarshal(signed.Signature, &rs); err != nil {
+			return fmt.Errorf("Failed to unmarshall ECDSA signature")
+		}
+		if !ecdsa.Verify(pubkey, hash[:], rs.R, rs.S) {
+			return errors.New("ECDSA signature validation failed")
+		}
+		return nil
+	default:
+		return fmt.Errorf("Unknown public key type %T", c.pubkey)
+	}
+
+}
+
+func (c *LogClient) verifySTH(sth *ct.SignedTreeHead) error {
+	if c.pubkey == nil {
+		// Can't verify signatures without the public key.
+		return nil
+	}
+
+	// RFC 6962 s3.5 specifies that the signature is over the following 50 bytes:
+	//  [0]     : 1-byte version = v1(0)
+	//  [1]     : 1-byte signature_type = tree_hash
+	//  [2:10]  : 8-byte timestamp
+	//  [10:18] : 8-byte tree size
+	//  [18:50] : 32-byte root hash
+	data := make([]byte, 50)
+	data[0] = byte(sth.Version)
+	data[1] = byte(ct.TreeHashSignatureType)
+	binary.BigEndian.PutUint64(data[2:], sth.Timestamp)
+	binary.BigEndian.PutUint64(data[10:], sth.TreeSize)
+	copy(data[18:], sth.SHA256RootHash[:])
+
+	return c.verifySignature(data, sth.TreeHeadSignature)
+}
+
+func (c *LogClient) verifySCT(sct *ct.SignedCertificateTimestamp, ctype ct.LogEntryType, certData ct.ASN1Cert) error {
```

Comment:
Might be cool to pull these two out into exported methods somewhere so that other things can use them too (maybe rename to `Verify{SCT,STH}Signature()` to be clear what they're doing?